gcc/cp/contracts.cc

```diff
@@ -1613,61 +1618,208 @@ build_contract_function_decls (tree d)
+	= FROB_CONTEXT (global_namespace);
+      TREE_PUBLIC (TYPE_NAME (pseudo_source_location_impl_type)) = true;
+      CLASSTYPE_LITERAL_P (pseudo_source_location_impl_type) = true;
+      CLASSTYPE_LAZY_COPY_CTOR (pseudo_source_location_impl_type) = true;
+      CLASSTYPE_LAZY_COPY_ASSIGN (pseudo_source_location_impl_type) = true;
+      CLASSTYPE_LAZY_DESTRUCTOR (pseudo_source_location_impl_type) = true;
+      xref_basetypes (pseudo_source_location_impl_type, /*bases=*/NULL_TREE);
+      pseudo_source_location_impl_type
+	= cp_build_qualified_type (pseudo_source_location_impl_type,
+				   TYPE_QUAL_CONST);
+    }
+
+  return pseudo_source_location_impl_type;
 }
+/* Build a layout-compatible internal version of source location type.  */
+static tree
+get_pseudo_source_location_type ()
+{
+  // build a source_location layout compatible type
+  if (!pseudo_source_location_type)
+    {
+      tree fields = NULL_TREE;
+      /* Must match <source_location>:
+       struct source_location
+	{
+	private:
+	    const __impl* _M_impl = nullptr;
+	};  */
+      const tree type = build_pointer_type (pseudo_source_location_impl_type);
 
-static const char *
-get_contract_role_name (tree contract)
+      tree next = build_decl (BUILTINS_LOCATION, FIELD_DECL,
+			      NULL_TREE, type);
+      DECL_CHAIN (next) = fields;
+      fields = next;
+
+      iloc_sentinel ils (input_location);
+      input_location = BUILTINS_LOCATION;
+      pseudo_source_location_type = make_class_type (RECORD_TYPE);
+      finish_builtin_struct (pseudo_source_location_type,
+			   "__pseudo_source_location_type",
+			   fields, NULL_TREE);
+      CLASSTYPE_AS_BASE (pseudo_source_location_type)
+      = pseudo_source_location_type;
+      DECL_CONTEXT (TYPE_NAME (pseudo_source_location_type))
+      = FROB_CONTEXT (global_namespace);
+      TREE_PUBLIC (TYPE_NAME (pseudo_source_location_type)) = true;
+      CLASSTYPE_LITERAL_P (pseudo_source_location_type) = true;
+      CLASSTYPE_LAZY_COPY_CTOR (pseudo_source_location_type) = true;
+      CLASSTYPE_LAZY_COPY_ASSIGN (pseudo_source_location_type) = true;
+      CLASSTYPE_LAZY_DESTRUCTOR (pseudo_source_location_type) = true;
+      xref_basetypes (pseudo_source_location_type, /*bases=*/NULL_TREE);
+      pseudo_source_location_type
+      = cp_build_qualified_type (pseudo_source_location_type,
+				 TYPE_QUAL_CONST);
+    }
+  return pseudo_source_location_type;
+}
+
+/* Build a layout-compatible internal version of source location __impl type.  */
+static tree
+build_pseudo_source_location_impl (tree contract)
 {
-  if (CONTRACT_LITERAL_MODE_P (contract))
-    return "";
-  if (tree mode = CONTRACT_MODE (contract))
-    if (tree role = TREE_PURPOSE (mode))
-      return IDENTIFIER_POINTER (role);
-  return "default";
+
+  expanded_location loc = expand_location (EXPR_LOCATION (contract));
+  const char *function = fndecl_name (DECL_ORIGIN (current_function_decl));
```

Comment:
I guess we are using DECL_ORIGIN() because we expect this to be called for helper functions? (this does use D_A_O).

Reply:
This is a direct copy paste from the previous way we got function names, I have not given it much thought. For the wrapper, it has the same problem that it prints out the wrapper name. At the moment, we have no way of recognising a wrapper function and getting to the name of the original function. This needs fixing. I've added an issue to the github